### PR TITLE
[SYCL] [AMDGPU] Ignore incorrect sub-group size

### PR DESCRIPTION
**Pull request by jchlanda**
CDNA supports only 64 wave front size, for those GPUs allow subgroup size of 64. Some GPUs support both 32 and 64, for those (and the rest) only allow 32.

### CODE REVIEW

**Comment by jchlanda** [comment]
Friendly ping @intel/dpcpp-cfe-reviewers

File: clang/test/SemaSYCL/reqd-sub-group-size-amd_64.cpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// RUN: %clang_cc1 -fsycl-is-device -triple amdgcn-amd-amdhsa -target-cpu gfx90a -internal-isystem %S/Inputs -std=c++2b -verify %s
+
+// Sub-group size is optimized for 32, warn (and ignore the attribute) if the
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/reqd-sub-group-size-amd_64.cpp:3]: This comment is incorrect for this test right?

Comment (jchlanda) [reply clang/test/SemaSYCL/reqd-sub-group-size-amd_64.cpp]: Yes, copy past error, fixed now.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -3526,6 +3526,10 @@ def warn_dllimport_dropped_from_inline_function : Warning<
 def warn_attribute_on_direct_kernel_callee_only : Warning<"%0 attribute allowed"
   " only on a function directly called from a SYCL kernel function; attribute ignored">,
   InGroup<IgnoredAttributes>;
+def warn_amd_reqd_sub_group_attribute_n
```
[the remark this answers is not in the document]
Comment (jchlanda) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:3529]: No problem, you are right we could fold it in and use the tirple name: 7076bdd4316cb15aefab99d75926f0deb40fd671

[the remark this answers is not in the document]
Comment (elizabethandrews) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:3544]: Did you make this change? The only thing I see in new commit is a change in test commentss

File: clang/test/SemaSYCL/reqd-sub-group-size-amd_32.cpp [hunk 3]
```diff
@@ -0,0 +1,24 @@
+// RUN: %clang_cc1 -fsycl-is-device -triple amdgcn-amd-amdhsa -target-cpu gfx1010 -internal-isystem %S/Inputs -std=c++2b -verify %s
+
+// Sub-group size is optimized for 32, warn (and ignore the attribute) if the
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/reqd-sub-group-size-amd_32.cpp:3]: FYI - You're not ignoring the attribute IIUC. It looks like you warn but still attach the attribute. I'm not sure how CodeGen handles this but as far as I can tell, the AST still has the attribute. You need to return early if you want to drop it.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 2 above)
[the remark this answers is not in the document]
Comment (jchlanda) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: Sorry, that's some git hiccup, should be done now.

**Comment by jchlanda** [comment]
Added a test to make sure that the values are indeed ignored: 0a72b7b

**Review by elizabethandrews — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]
LGTM

**Comment by againull** [comment]
Test failure:
  SYCL :: OptionalKernelFeatures/is_compatible/is_compatible_with_aspects.cpp
seems to be related to the patch.

**Comment by elizabethandrews** [comment]
> way this test works is that it [creates a kernel with a known, incorrect required sub gr](https://github.com/jchlanda/llvm/blob/jakub/launch_bounds_warnings/sycl/test-e2e/OptionalKernelFeatures/is_compatible/is_compatible_with_aspects.cpp#L75)

I don't think it makes sense allowing an invalid subgroup size to pass through if we're emitting the warning.  IMO we should either just allow the incorrect value to pass through without a warning and then have the backend deal with it however it is doing so currently,  or we do what this PR does - i.e. emit warning and drop the attribute. @AlexeySachkov can you weigh in here? Is there a reason we are passing though invalid subgroup size?

**Comment by jchlanda** [comment]
Any preference on this @AlexeySachkov ?

**Comment by jchlanda** [comment]
> Any preference on this @AlexeySachkov ?

@AlexeySachkov friendly ping on this one please.

**Comment by AlexeySachkov** [comment]
> > > Test failure: SYCL :: OptionalKernelFeatures/is_compatible/is_compatible_with_aspects.cpp seems to be related to the patch.
> > 
> > 
> > That's right, this patch changes the handling of incorrect sub-group attribute, such that it doesn't end up in the resulting binary image. The way this test works is that it [creates a kernel with a known, incorrect required sub group size (INT_MAX)](https://github.com/jchlanda/llvm/blob/jakub/launch_bounds_warnings/sycl/test-e2e/OptionalKernelFeatures/is_compatible/is_compatible_with_aspects.cpp#L75). Then expects that kernel to be incompatible with the device, by [failing a check against supported sub-group sizes](https://github.com/jchlanda/llvm/blob/jakub/launch_bounds_warnings/sycl/source/detail/program_manager/program_manager.cpp#L2848).
> > There are two ways of thinking about that issue:
> > 
> > * since we warn and ignore the incorrect value of sub-group size: https://github.com/intel/llvm/pull/11687/files#diff-2a5bdb2d9f07f8d77de51d5403d349c22978141b6de6bd87fc5e449f5ed95becR4027 this test is not applicable anymore,
> > * or we shouldn't ignore the incorrect values and only warn.
> > 
> > I'm not sure which is the correct way of handling those values. Would you be able to advise @againull @elizabethandrews @smanna12 ?
> 
> I don't think it makes sense allowing an invalid subgroup size to pass through if we're emitting the warning. IMO we should either just allow the incorrect value to pass through without a warning and then have the backend deal with it however it is doing so currently, or we do what this PR does - i.e. emit warning and drop the attribute. @AlexeySachkov can you weigh in here? Is there a reason we are passing though invalid subgroup size?

From SYCL spec point of view, there are no compile-time known incorrect sub-group sizes - what we have for AMD and CUDA are implementation details of those backends. 

Explicitly requesting sub-group size of a kernel has limited portability by design, but considering how fundamental and narrow the selection of possible sub-group sizes for AMD/CUDA, it might make sense to promote that knowledge into the compiler in form of the suggested warning to improve user experience. 

I think that it may actually make sense to pass the invalid sizes through even if we diagnosed them with a warning: this way runtime handling of the attribute for AMD/CUDA backend would be uniform with other backends and users will get more consistent experience. At the same time the warning will let users know in advance that their application won't work and it requires some code changes.

The only concern about the warning I have is that it may produce false alarms: what if there is a kernel which is never submitted to a AMD/CUDA device? The warning will still be there and it may be annoying for someone who uses `-Werror`. Once we get proper support for optional kernel features for AOT targets it may be even easier to discover that

**Comment by jchlanda** [comment]
> > > > Test failure: SYCL :: OptionalKernelFeatures/is_compatible/is_compatible_with_aspects.cpp seems to be related to the patch.
> > > 
> > > 
> > > That's right, this patch changes the handling of incorrect sub-group attribute, such that it doesn't end up in the resulting binary image. The way this test works is that it [creates a kernel with a known, incorrect required sub group size (INT_MAX)](https://github.com/jchlanda/llvm/blob/jakub/launch_bounds_warnings/sycl/test-e2e/OptionalKernelFeatures/is_compatible/is_compatible_with_aspects.cpp#L75). Then expects that kernel to be incompatible with the device, by [failing a check against supported sub-group sizes](https://github.com/jchlanda/llvm/blob/jakub/launch_bounds_warnings/sycl/source/detail/program_manager/program_manager.cpp#L2848).
> > > There are two ways of thinking about that issue:
> > > 
> > > * since we warn and ignore the incorrect value of sub-group size: https://github.com/intel/llvm/pull/11687/files#diff-2a5bdb2d9f07f8d77de51d5403d349c22978141b6de6bd87fc5e449f5ed95becR4027 this test is not applicable anymore,
> > > * or we shouldn't ignore the incorrect values and only warn.
> > > 
> > > I'm not sure which is the correct way of handling those values. Would you be able to advise @againull @elizabethandrews @smanna12 ?
> > 
> > 
> > I don't think it makes sense allowing an invalid subgroup size to pass through if we're emitting the warning. IMO we should either just allow the incorrect value to pass through without a warning and then have the backend deal with it however it is doing so currently, or we do what this PR does - i.e. emit warning and drop the attribute. @AlexeySachkov can you weigh in here? Is there a reason we are passing though invalid subgroup size?
> 
> From SYCL spec point of view, there are no compile-time known incorrect sub-group sizes - what we have for AMD and CUDA are implementation details of those backends.
> 
> Explicitly requesting sub-group size of a kernel has limited portability by design, but considering how fundamental and narrow the selection of possible sub-group sizes for AMD/CUDA, it might make sense to promote that knowledge into the compiler in form of the suggested warning to improve user experience.
> 
> I think that it may actually make sense to pass the invalid sizes through even if we diagnosed them with a warning: this way runtime handling of the attribute for AMD/CUDA backend would be uniform with other backends and users will get more consistent experience. At the same time the warning will let users know in advance that their application won't work and it requires some code changes.
> 

Following you suggestion, I've let the invalid value pass through. I've removed a test, since no values are being ignored.

> The only concern about the warning I have is that it may produce false alarms: what if there is a kernel which is never submitted to a AMD/CUDA device? The warning will still be there and it may be annoying for someone who uses `-Werror`. Once we get proper support for optional kernel features for AOT targets it may be even easier to discover that

I feel that it's an edge case, that is outweighed by the benefit of informing the users about the incorrect size, and perhaps we could ignore it for now. If that ever becomes an issue, it would be simple enough to provide a mechanism to silence this particular warning?

**Comment by AlexeySachkov** [comment]
> > The only concern about the warning I have is that it may produce false alarms: what if there is a kernel which is never submitted to a AMD/CUDA device? The warning will still be there and it may be annoying for someone who uses `-Werror`. Once we get proper support for optional kernel features for AOT targets it may be even easier to discover that
> 
> I feel that it's an edge case, that is outweighed by the benefit of informing the users about the incorrect size, and perhaps we could ignore it for now. If that ever becomes an issue, it would be simple enough to provide a mechanism to silence this particular warning?

I think that we may encounter that sooner than later, but I do agree that benefit of the warning probably outweighs the `-Werror` concern. I'm fine with the direction of this PR, but we should keep in mind that at some point we will likely need a know to disable that particular warning.

**Comment by al42and** [comment]
> I feel that it's an edge case, that is outweighed by the benefit of informing the users about the incorrect size, and perhaps we could ignore it for now. If that ever becomes an issue, it would be simple enough to provide a mechanism to silence this particular warning?

As a person compiling with `-Werror` and for multiple sub-group sizes, I mostly agree there.

It looks like currently the required error suppression would be too broad. Originally (#6183), `-Wno-cuda-compat` was enough, which does not suppress too much relevant warnings. Now, `-Wno-unknown-attributes` would be needed, which can hide true positives.

**Pull request merged**